llvm/test/Feature/Repo/repo_pruning_dependents.py

```diff
@@ -0,0 +1,306 @@
+#
+# In the llvm-project-prepo project, the GO's digest could be generated either by the
+# RepoMeatadataGenerator pass or by the RepoObjectWriter. When a new global object (GO) is
+# generated by an LLVM optimisation pass after the RepoMeatadataGenerator pass, the digest
+# will be generated by the RepoObjectWriter. If a fragment has an external fixup
+# referencing this GO, we add the GO's compilation member address to the dependents.
+#
+# When a fragment is pruned, all its dependents need to be recorded in the repo.tickets
+# metadata to guarantee that the dependents (those GO generated by later optimizations)
+# are present in the final compilation record.
+#
+# ----  Test1 ----
+# In the test1, the digest of 'A' is given in the TicketNode (as if it were generated
+# by the RepoMeatadataGenerator pass). The digests of `B` and `C` will be generated by the
+# RepoObjectWriter. Since Function 'A' is dependent on function 'B' and 'B' is dependent
+# on function 'C', both TicketNodes of 'B' and 'C' need to be added into the 'repo.tickets'
+# during the pruning.
+#
+# RUN: rm -f %t.db
+# Step 1: write a test1 IR code targeted on ELF.
+# RUN: python %s -t 1 -o %t_1.elf.ll
+# Step 2: generate a test1 IR code targeted on Repo.
+# RUN: env REPOFILE=%t.db clang -O0 -S -emit-llvm --target=x86_64-pc-linux-gnu-repo -x ir %t_1.elf.ll  -o %t_1.repo.ll
+# Step 3: remove the TicketNode metadata from the functions B and C.
+# RUN: python %s -f B C -i %t_1.repo.ll -o %t_1.ll
+# Step 4: Since the functions B and C don't have the TicketNode, the digest of functions `B` and `C` are generated
+#         by the RepoObjectWriter and added these two functions into the dependents of the function 'A'.
+# RUN: env REPOFILE=%t.db llc -mtriple="x86_64-pc-linux-gnu-repo" -filetype=obj %t_1.ll -o %t_1.1.o
+# Step 5: Dump all compilation members in the database.
+#         There are 3 compilation members, which are functions `A`, `B` and `C`.
+# RUN: env REPOFILE=%t.db pstore-dump  --all-compilations %t.db > %t_1.1.log
+# Step 6: Re-build the same ELF IR code again.
+#         In this step, the IR code firstly pass the RepoMeatadataGenerator pass to generate the
+#         digests of functions 'A', 'B' and 'C'. During the RepoPruning pass, the function 'A'
+#         is pruned and its dependents 'B' and 'C' are added.
+# RUN: env REPOFILE=%t.db clang -O0 -c --target=x86_64-pc-linux-gnu-repo -x ir %t_1.elf.ll -o %t_1.2.o
+# Step 7: Dump all compilation members in the database again.
+#         The compilations, which are generated by step 4 and step 6, should be the same.
+#         Therefore, there are only one compilation, which includes three members: functions
+#          `A`, `B` and `C`.
+# RUN: env REPOFILE=%t.db pstore-dump  --all-compilations %t.db > %t_1.2.log
+# Step 8: Check that the step 5 and step 7 generate the same compilation members.
+# RUN: diff %t_1.1.log %t_1.2.log
+
+import argparse
+import re
+
+def write_elf_IR_test1 (filename):
```

Comment:
Could these be files in the Input directory?

Reply:
In the LLVM test, the python test normally doesn’t mix with the IR code. I went through some LLVM python tests. The python file writes the test and run the check on the generated test. Therefore, the test file is not put into the Input directory.

Reply:
As it is now, this is a big improvement over the tests which needed the digest updating manually. I do find the IR harder to read in a series of f.write statements though, and that obscures the aim of the test(s).

I had a look through some LLVM tests and I think that there are two "styles" of how python is used in tests (there may be more):
1). e.g. llvm/test/CodeGen/SystemZ/Large/branch-range-01.py which uses python to generate IR, and needs to because there is a process to generate the IR (with loops for example). In these repo tests the IR is fixed so I don't think that we need to produce it from python based on this kind of test.
2). e.g. llvm/test/tools/llvm-objcopy/Inputs/ungzip.py which puts some utility python scripts in "Inputs" and uses those across several tests.

Would 2) be a better fit for this? We don't need to generate complicated IR, we just need a bit more functionality than we can do with lit and FileCheck. remove_function_ticketnode() and remove_variable_ticketnode() could be in python scripts called with (e.g.) "%python %S/Inputs/remove_function_ticketnode.py" and the IR could be kept in separate .ll files.

Reply:
Good point. Updated the tests following your advice.